File: apps/api/test/template.test.ts
import { describe, it, expect } from "vitest";
import { renderTemplate, contactVars } from "../src/modules/campaigns/routes.js";

describe("renderTemplate (blasting variables)", () => {
  it("mengganti variabel yang dikenal", () => {
    expect(
      renderTemplate("Halo {{nama}}, dari {{perusahaan}}!", { nama: "Budi", perusahaan: "PT X" }),
    ).toBe("Halo Budi, dari PT X!");
  });

  it("case-insensitive & toleran spasi", () => {
    expect(renderTemplate("Hi {{ Nama }}!", { nama: "Siti" })).toBe("Hi Siti!");
  });

  it("variabel tak dikenal dibiarkan (typo mudah terdeteksi)", () => {
    expect(renderTemplate("Halo {{namma}}", { nama: "Budi" })).toBe("Halo {{namma}}");
  });

  it("contactVars memetakan field kontak", () => {
    const v = contactVars({ fullName: "Budi Santoso", companyName: "PT X", email: "b@x.id", phone: "+6281" });
    expect(v.nama).toBe("Budi");
    expect(v.nama_lengkap).toBe("Budi Santoso");
    expect(v.perusahaan).toBe("PT X");
  });
});

File: internal/wa/template_test.go
package wa

import "testing"

func TestRenderTemplate(t *testing.T) {
	got := RenderTemplate("Halo {{nama}}, dari {{perusahaan}}!", map[string]string{"nama": "Budi", "perusahaan": "PT X"})
	if got != "Halo Budi, dari PT X!" {
		t.Fatalf("got %q", got)
	}
	if got := RenderTemplate("Hi {{ Nama }}!", map[string]string{"nama": "Siti"}); got != "Hi Siti!" {
		t.Fatalf("case-insensitive gagal: %q", got)
	}
	if got := RenderTemplate("Halo {{namma}}", map[string]string{"nama": "Budi"}); got != "Halo {{namma}}" {
		t.Fatalf("unknown key harus dibiarkan: %q", got)
	}
	v := ContactVars("Budi Santoso", "PT X", "b@x.id", "+6281")
	if v["nama"] != "Budi" || v["nama_lengkap"] != "Budi Santoso" {
		t.Fatalf("contact vars salah: %v", v)
	}
}
